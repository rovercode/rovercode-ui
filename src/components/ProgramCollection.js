import React, { Component } from 'react';
import { Link } from 'react-router-dom';
import { FormattedMessage, injectIntl } from 'react-intl';
import PropTypes from 'prop-types';
import { hot } from 'react-hot-loader';
import {
  Box,
  Container,
  Grid,
  Button,
  Typography,
  Chip,
  TextField,
  InputAdornment,
  Menu,
  MenuItem,
  Card,
  CardContent,
  CardActionArea,
  CardActions,
  Link as MuiLink,
} from '@material-ui/core';
import { grey } from '@material-ui/core/colors';
import { withStyles } from '@material-ui/core/styles';
import {
  Add,
  ArrowDownward,
  ArrowUpward,
  Search,
  Delete,
} from '@material-ui/icons';
import { Pagination, Autocomplete } from '@material-ui/lab';
import { debounce } from 'throttle-debounce';

import flavourImage from '@/assets/images/flavour.png';

const styles = (theme) => ({
  mainContainer: {
    marginBottom: theme.spacing(8),
    [theme.breakpoints.up('xs')]: {
      minWidth: theme.breakpoints.values.xs,
    },
    [theme.breakpoints.up('sm')]: {
      minWidth: theme.breakpoints.values.sm,
    },
    [theme.breakpoints.up('md')]: {
      minWidth: theme.breakpoints.values.md,
    },
    [theme.breakpoints.up('lg')]: {
      minWidth: theme.breakpoints.values.lg,
    },
  },
  paginationPaddedBox: {
    display: 'flex',
    justifyContent: 'center',
    marginTop: theme.spacing(3),
    marginBottom: theme.spacing(3),
  },
});

class ProgramCollection extends Component {
  constructor(props) {
    super(props);

    this.searchDebounce = debounce(
      parseInt(SEARCH_DEBOUNCE_TIME, 10) || 1000, this.update, // eslint-disable-line no-undef
    );

    this.state = {
      page: 1,
      searchQuery: null,
      ordering: 'name',
      tagFilters: [],
      sortMenuAnchorElement: null,
    };
  }

  update = () => {
    const { onUpdate } = this.props;
    const {
      page,
      ordering,
      searchQuery,
      tagFilters,
    } = this.state;

    const params = {
      page,
      ordering,
      tag: tagFilters.join(),
    };

    if (searchQuery) {
      params.search = searchQuery;
    }

    onUpdate(params);
  }

  toggleOrdering = (name) => {
    const { ordering } = this.state;

    if (ordering.endsWith(name)) {
      if (ordering.startsWith('-')) {
        return ordering.substr(1);
      }

      return `-${ordering}`;
    }

    return name;
  }

  handleSortClick = (event) => this.setState({
    sortMenuAnchorElement: event.currentTarget,
  })

  handleSortClose = () => this.setState({
    sortMenuAnchorElement: null,
  })

  handlePageChange = (e, page) => this.setState({
    page,
  }, this.update)

  handleSearchChange = (event) => this.setState({
    searchQuery: event.target.value,
    page: 1,
  }, this.searchDebounce)

  handleOrderingChange = (e) => this.setState({
    ordering: this.toggleOrdering(e.target.id),
  }, this.update)

  handleTagFilterChange = (event, value) => this.setState({
    tagFilters: value,
  }, this.update)

  render() {
    const {
      user,
      label,
      onProgramClick,
      onRemoveClick,
      programs,
      intl,
      tag,
      classes,
    } = this.props;
    const {
      ordering,
      searchQuery,
      sortMenuAnchorElement,
      tagFilters,
    } = this.state;

    const searchPlaceholder = intl.formatMessage({
      id: 'app.program_collection.search',
      description: 'Placeholder for search entry',
      defaultMessage: 'Search programs',
    });

    const tagFilterPlaceholder = intl.formatMessage({
      id: 'app.program_collection.filter',
      description: 'Placeholder for selecting tags used to filter',
      defaultMessage: 'Filter by tag',
    });

    const Title = withStyles(() => ({
      root: {
        fontWeight: 'bold',
      },
    }))(Typography);

    const DeleteButton = withStyles(() => ({
      root: {
        color: grey[500],
      },
    }))(Button);

    const FixedWidthAutocomplete = withStyles(() => ({
      root: {
        width: '100%',
      },
    }))(Autocomplete);

    const TitleArea = withStyles((theme) => ({
      root: {
        marginBottom: theme.spacing(4),
      },
    }))(Grid);

    let programCount = -1;
    let programLimit = -1;
    if (user.stats && user.stats.block_diagram) {
      programCount = user.stats.block_diagram.count;
      programLimit = user.stats.block_diagram.limit;
    }

    return (
      <Container className={classes.mainContainer}>
        <TitleArea item container direction="row" justify="space-between">
          <Grid item>
            <Title variant="h4">
              {label}
            </Title>
          </Grid>
          <Grid item>
            <Button
              variant="contained"
              color="primary"
              size="large"
              startIcon={<Add />}
              component={Link}
              to="/mission-control"
              disabled={user.tier === 1 && programCount >= programLimit}
            >
              <FormattedMessage
                id="app.program_collection.new"
                description="Button label to create new program"
                defaultMessage="New Program"
              />
            </Button>
          </Grid>
        </TitleArea>
        <Grid container spacing={2}>
          <Grid item xs={4}>
            <TextField
              variant="outlined"
              size="small"
              className="prompt"
              icon="search"
              placeholder={searchPlaceholder}
              onChange={this.handleSearchChange}
              InputProps={{
                endAdornment: (
                  <InputAdornment position="end">
                    <Search />
                  </InputAdornment>
                ),
              }}
            />
          </Grid>
          <Grid item xs={8}>
            <Grid container justify="flex-end" spacing={2}>
              <Grid item>
                <Button aria-controls="sort-menu" aria-haspopup="true" onClick={this.handleSortClick}>
                  <FormattedMessage
                    id="app.program_collection.sort"
                    description="Button label for sort options"
                    defaultMessage="Sort"
                  />
                </Button>
                <Menu
                  id="sort-menu"
                  anchorEl={sortMenuAnchorElement}
                  keepMounted
                  open={Boolean(sortMenuAnchorElement)}
                  onClose={this.handleSortClose}
                  getContentAnchorEl={null}
                  anchorOrigin={{
                    vertical: 'bottom',
                    horizontal: 'center',
                  }}
                  transformOrigin={{
                    vertical: 'top',
                    horizontal: 'center',
                  }}
                >
                  <MenuItem onClick={this.handleOrderingChange} id="name">
                    <FormattedMessage
                      id="app.program_collection.name"
                      description="Button label to sort by name"
                      defaultMessage="Name"
                    />
                    {
                      ordering === 'name' ? (
                        <>
                          <ArrowDownward />
                        </>
                      ) : (
                        <>
                          <ArrowUpward />
                        </>
                      )
                    }
                  </MenuItem>
                </Menu>
              </Grid>
              <Grid item xs={2}>
                <FixedWidthAutocomplete
                  id="tag-select"
                  multiple
                  freeSolo
                  filterSelectedOptions
                  size="small"
                  options={tag.tags.map((t) => t.name)}
                  onChange={this.handleTagFilterChange}
                  value={tagFilters}
                  renderInput={(params) => (
                    <TextField {...params} variant="outlined" label={tagFilterPlaceholder} />
                  )}
                  renderTags={(value, getTagProps) => value.map((option, index) => (
                    <Chip color="secondary" size="small" label={option} {...getTagProps({ index })} />
                  ))}
                />
              </Grid>
            </Grid>
          </Grid>
        </Grid>
        {
          user.tier === 1 && programCount >= 0 ? (
            <Grid container direction="column" justify="flex-start" spacing={2}>
              <Grid item>
                <Typography variant="h5" display="inline">
                  <FormattedMessage
                    id="app.program_collection.slots"
                    description="Lists the number of free slots"
                    defaultMessage="Free Program Slots"
                  />
                  {': '}
                </Typography>
                <Typography variant="h5" color="secondary" display="inline">
                  {`${programCount}/${programLimit}`}
                </Typography>
              </Grid>
              {
                programCount >= programLimit ? (
                  <Grid item>
                    <Typography variant="subtitle1">
                      <FormattedMessage
                        id="app.program_collection.over1"
                        description="Notifies the user of no remaining free slots"
                        defaultMessage="You have used all your free program slots. You can delete an existing program to free up a program slot, or you can"
                      />
                      {' '}
                      <MuiLink href="/user/settings">
                        <FormattedMessage
                          id="app.program_collection.over2"
                          description="Notifies the user of no remaining free slots"
                          defaultMessage="upgrade your account"
                        />
                      </MuiLink>
                      {' '}
                      <FormattedMessage
                        id="app.program_collection.over3"
                        description="Notifies the user of no remaining free slots"
                        defaultMessage="for unlimited programs."
                      />
                    </Typography>
                  </Grid>
                ) : (null)
              }
            </Grid>
          ) : (null)
        }
        <Grid container spacing={3}>
          {
            programs.count === 0 && !searchQuery && tagFilters.length === 0 ? (
              <Grid item container direction="column" alignItems="center" justify="center">
                <Grid item>
                  <img alt="Kids" width="300px" src={flavourImage} />
                </Grid>
                <Grid item>
                  <Typography variant="h4">
                    <FormattedMessage
                      id="app.program_collection.no_programs"
                      description="Notifies new users he or she has no programs"
                      defaultMessage="You don't have any programs yet!"
                    />
                  </Typography>
                </Grid>
              </Grid>
            ) : (null)
          }
          {
            programs.count === 0 && (searchQuery || tagFilters.length !== 0) ? (
              <Grid item container direction="row" alignItems="center" justify="center">
                <Typography variant="h4">
                  <FormattedMessage
                    id="app.program_collection.nothing"
                    description="Informs the user that no programs match the filters"
                    defaultMessage="Sorry, no programs match your filters."
                  />
                </Typography>
              </Grid>
            ) : programs.results.map((program) => (
              <Grid item xs={12} md={6} lg={3} key={program.id}>
                <Card key={program.id}>
                  <CardActionArea
                    id={program.id}
                    data-owned={user.username === program.user.username}
                    onClick={onProgramClick}
                  >
                    <CardContent>
                      <Typography variant="h6">
                        {program.name}
                      </Typography>
                      {
                        user.username === program.user.username ? null : (
                          <Typography variant="subtitle2">
                            {program.user.username}
                          </Typography>
                        )
                      }
                    </CardContent>
                  </CardActionArea>
                  {
                    user.username === program.user.username ? (
                      <CardActions>
                        <DeleteButton
                          id={program.id}
                          name={program.name}
                          onClick={onRemoveClick}
                          size="small"
                          startIcon={<Delete />}
                        >
                          <FormattedMessage
                            id="app.program_collection.remove"
                            description="Button label to remove a program"
                            defaultMessage="Delete"
                          />
                        </DeleteButton>
                      </CardActions>
                    ) : (null)
                  }
                </Card>
              </Grid>
            ))
          }
        </Grid>
        {
          programs.total_pages > 1 ? (
            <Box
              className={classes.paginationPaddedBox}
            >
              <Pagination
                defaultPage={1}
                count={programs.total_pages}
                showFirstButton
                showLastButton
                color="secondary"
                onChange={this.handlePageChange}
              />
            </Box>
          ) : (null)
        }
      </Container>
    );
  }
}

ProgramCollection.defaultProps = {
  tag: {
    tags: [],
  },
};

ProgramCollection.propTypes = {
  classes: PropTypes.shape({
    mainContainer: PropTypes.string.isRequired,
    paginationPaddedBox: PropTypes.string.isRequired,
  }).isRequired,
  user: PropTypes.shape({
    username: PropTypes.string.isRequired,
    tier: PropTypes.number.isRequired,
    stats: PropTypes.shape({
      block_diagram: PropTypes.shape({
        count: PropTypes.number,
        limit: PropTypes.number,
      }),
    }),
  }).isRequired,
  programs: PropTypes.shape({
    count: PropTypes.number,
    next: PropTypes.string,
    previous: PropTypes.string,
    total_pages: PropTypes.number,
    free_max: PropTypes.number,
    results: PropTypes.arrayOf(
      PropTypes.shape({
        id: PropTypes.number.isRequired,
        name: PropTypes.string.isRequired,
        user: PropTypes.shape({
          username: PropTypes.string.isRequired,
        }).isRequired,
      }),
    ),
  }).isRequired,
  tag: PropTypes.shape({
    tags: PropTypes.arrayOf(PropTypes.shape({
      name: PropTypes.string,
    })),
  }),
  label: PropTypes.string.isRequired,
  onProgramClick: PropTypes.func.isRequired,
  onRemoveClick: PropTypes.func.isRequired,
  onUpdate: PropTypes.func.isRequired,
  intl: PropTypes.shape({
    formatMessage: PropTypes.func.isRequired,
  }).isRequired,
};

export default hot(module)(injectIntl(withStyles(styles)(ProgramCollection)));
