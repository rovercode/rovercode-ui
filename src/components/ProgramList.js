import React, { Component } from 'react';
import {
  Box,
  CircularProgress,
  Grid,
  Divider,
} from '@material-ui/core';
import { injectIntl } from 'react-intl';
import { Redirect } from 'react-router-dom';
import PropTypes from 'prop-types';

import ConfirmDialog from './ConfirmDialog';
import Footer from './Footer';
import ProgramCollection from './ProgramCollection';

const defaultState = {
  programSelected: null,
  confirmOpen: false,
  focusProgram: {
    id: null,
    name: null,
  },
};

class ProgramList extends Component {
  constructor(props) {
    super(props);

    this.state = defaultState;
  }

  componentDidMount() {
    const { clearProgram, clearProgramList } = this.props;

    clearProgramList();
    clearProgram();

    return this.fetch();
  }

  fetch = () => {
    const {
      fetchPrograms,
      fetchTags,
      fetchUserStats,
      owned,
      user,
    } = this.props;

    fetchTags();
    fetchUserStats(user.user_id);

    return owned ? fetchPrograms({
      user: user.user_id,
    }) : fetchPrograms({
      user__not: user.user_id,
    });
  }

  showConfirm = (e) => this.setState({
    confirmOpen: true,
    focusProgram: {
      id: e.target.parentNode.id || e.target.id,
      name: e.target.parentNode.name || e.target.name,
    },
  })

  cancelRemove = () => this.setState(defaultState)

  removeProgram = () => {
    const { removeProgram } = this.props;
    const { focusProgram } = this.state;

    this.setState(defaultState);

    return removeProgram(focusProgram.id).then(this.fetch);
  }

  selectProgram = (e) => {
    let program = e.target;
    if (e.target.parentNode.parentNode.id) {
      program = e.target.parentNode.parentNode;
    }

    this.setState({
      programSelected: program.id,
    });
  }

  fetchUserPrograms = (params) => {
    const { fetchPrograms, user } = this.props;

    fetchPrograms({
      user: user.user_id,
      ...params,
    });
  }

  fetchCommunityPrograms = (params) => {
    const { fetchPrograms, user } = this.props;
    fetchPrograms({
      user__not: user.user_id,
      ...params,
    });
  }

  render() {
    const {
      intl,
      owned,
      programs,
      user,
      tag,
    } = this.props;
    const {
      confirmOpen,
      focusProgram,
      programSelected,
    } = this.state;

    const myProgramsHeader = intl.formatMessage({
      id: 'app.program_list.my_programs',
      description: 'Header for all of user\'s programs',
      defaultMessage: 'My Programs',
    });

    const communityProgramsHeader = intl.formatMessage({
      id: 'app.program_list.community_programs',
      description: 'Header for all community programs',
      defaultMessage: 'Community Programs',
    });

    const cancelButtonText = intl.formatMessage({
      id: 'app.program_list.cancel',
      description: 'Button label to cancel removing program',
      defaultMessage: 'No, don\'t remove program',
    });

    const confirmButtonText = intl.formatMessage({
      id: 'app.program_list.confirm',
      description: 'Button label to confirm removing program',
      defaultMessage: 'Yes, remove program',
    });

    const dialogHeader = intl.formatMessage({
      id: 'app.program_list.dialog_header',
      description: 'Header for removing program confirmation dialog',
      defaultMessage: 'Remove Program',
    });

    const dialogContent = intl.formatMessage({
      id: 'app.program_list.dialog_content',
      description: 'Asks the user to confirm removing program',
      defaultMessage: 'Are you sure you want to remove {name}?',
    }, {
      name: focusProgram.name,
    });

    return (
      <>
        {
          programSelected ? (
            <Redirect to={{
              pathname: `/mission-control/${programSelected}`,
            }}
            />
          ) : (null)
        }
        <Grid container direction="column" justify="center" alignItems="center">
          {
          programs === null ? (
            <Grid item>
              <CircularProgress />
            </Grid>
          ) : (
            <Grid item>
              <Box m={2}>
                <ProgramCollection
                  programs={programs}
                  user={user}
                  tag={tag}
                  label={owned ? myProgramsHeader : communityProgramsHeader}
                  onProgramClick={this.selectProgram}
                  onRemoveClick={this.showConfirm}
                  onUpdate={owned ? this.fetchUserPrograms : this.fetchCommunityPrograms}
                />
              </Box>
            </Grid>
          )
        }
        </Grid>
        <ConfirmDialog
          title={dialogHeader}
          open={confirmOpen}
          onConfirm={this.removeProgram}
          onCancel={this.cancelRemove}
          cancelButton={cancelButtonText}
          confirmButton={confirmButtonText}
        >
          {dialogContent}
        </ConfirmDialog>
        <Divider />
        <Footer />
      </>
    );
  }
}

ProgramList.defaultProps = {
  programs: {
    next: null,
    previous: null,
    total_pages: 1,
    results: [],
  },
  tag: {
    tags: [],
  },
};

ProgramList.propTypes = {
  fetchPrograms: PropTypes.func.isRequired,
  fetchTags: PropTypes.func.isRequired,
  removeProgram: PropTypes.func.isRequired,
  clearProgramList: PropTypes.func.isRequired,
  clearProgram: PropTypes.func.isRequired,
  fetchUserStats: PropTypes.func.isRequired,
  owned: PropTypes.bool.isRequired,
  user: PropTypes.shape({
    user_id: PropTypes.number.isRequired,
    username: PropTypes.string.isRequired,
  }).isRequired,
  programs: PropTypes.shape({
    next: PropTypes.string,
    previous: PropTypes.string,
    total_pages: PropTypes.number,
    results: PropTypes.arrayOf(
      PropTypes.shape({
        id: PropTypes.number.isRequired,
        name: PropTypes.string.isRequired,
        user: PropTypes.shape({
          username: PropTypes.string.isRequired,
        }).isRequired,
      }),
    ),
  }),
  tag: PropTypes.shape({
    tags: PropTypes.arrayOf(PropTypes.shape({
      name: PropTypes.string,
    })),
  }),
  intl: PropTypes.shape({
    formatMessage: PropTypes.func.isRequired,
  }).isRequired,
};

export default injectIntl(ProgramList);
