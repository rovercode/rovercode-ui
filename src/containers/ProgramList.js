import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { hot } from 'react-hot-loader';
import { withCookies, Cookies } from 'react-cookie';
import { changeReadOnly as actionChangeReadOnly, clearProgram } from '../actions/code';
import { clearPrograms, fetchPrograms, removeProgram } from '../actions/program';
import { checkAuthError, authHeader } from '../actions/auth';
import { fetchTags } from '../actions/tag';
import { fetchUserStats } from '../actions/user';
import ProgramList from '../components/ProgramList';

const mapStateToProps = ({
  code, program, tag, user,
}) => ({
  code, ...program, tag, user,
});
const mapDispatchToProps = (dispatch, { cookies }) => ({
  fetchPrograms: (params) => {
    const xhrOptions = authHeader(cookies);

    xhrOptions.params = params;

    return dispatch(fetchPrograms(xhrOptions)).catch(checkAuthError(dispatch));
  },
  removeProgram: (id) => dispatch(removeProgram(id, authHeader(cookies)))
    .catch(checkAuthError(dispatch)),
  changeReadOnly: (isReadOnly) => dispatch(actionChangeReadOnly(isReadOnly)),
  fetchTags: () => dispatch(fetchTags(authHeader(cookies))).catch(checkAuthError(dispatch)),
  clearProgram: () => dispatch(clearProgram()),
  clearProgramList: () => dispatch(clearPrograms()),
  fetchUserStats: (id) => dispatch(fetchUserStats(id, authHeader(cookies)))
    .catch(checkAuthError(dispatch)),
});

const ProgramListContainer = connect(
  mapStateToProps,
  mapDispatchToProps,
)(ProgramList);

ProgramListContainer.defaultProps = {
  owned: false,
};

ProgramListContainer.propTypes = {
  cookies: PropTypes.instanceOf(Cookies).isRequired,
  owned: PropTypes.bool,
};

export default hot(module)(withCookies(ProgramListContainer));
