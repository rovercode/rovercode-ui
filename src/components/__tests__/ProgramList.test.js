import React from 'react';
import { MemoryRouter } from 'react-router';
import { CircularProgress } from '@material-ui/core';
import { Redirect } from 'react-router-dom';
import ProgramCollection from '../ProgramCollection';
import ProgramList from '../ProgramList';

let fetchPrograms;
let removeProgram;
let fetchTags;
let clearProgram;
let clearProgramList;
let fetchUserStats;

describe('The ProgramList component', () => {
  beforeEach(() => {
    fetchPrograms = jest.fn().mockResolvedValue();
    removeProgram = jest.fn().mockResolvedValue();
    fetchTags = jest.fn().mockResolvedValue();
    clearProgram = jest.fn();
    clearProgramList = jest.fn();
    fetchUserStats = jest.fn().mockResolvedValue();
  });

  test('renders on the page with no errors', () => {
    const wrapper = shallowWithIntl(
      <ProgramList
        fetchPrograms={fetchPrograms}
        removeProgram={removeProgram}
        fetchTags={fetchTags}
        clearProgram={clearProgram}
        clearProgramList={clearProgramList}
        fetchUserStats={fetchUserStats}
        user={{ user_id: 1, username: 'testuser', tier: 1 }}
        owned
      />,
    ).dive().dive();
    wrapper.setState({
      focusProgram: {
        id: 1,
        name: 'Test Program',
      },
    });
    expect(wrapper).toMatchSnapshot();
  });

  test('fetches programs and tags on mount', async () => {
    await mountWithIntl(
      <MemoryRouter>
        <ProgramList
          fetchPrograms={fetchPrograms}
          removeProgram={removeProgram}
          fetchTags={fetchTags}
          clearProgram={clearProgram}
          clearProgramList={clearProgramList}
          fetchUserStats={fetchUserStats}
          user={{ user_id: 1, username: 'testuser', tier: 1 }}
          owned
        />
      </MemoryRouter>,
    );
    expect(fetchTags.mock.calls.length).toBe(1);
    expect(fetchPrograms.mock.calls.length).toBe(1);
    expect(clearProgram.mock.calls.length).toBe(1);
  });

  test('shows the correct number of programs for the user', () => {
    const programs = {
      next: null,
      previous: null,
      total_pages: 1,
      results: [{
        id: 33,
        name: 'Unnamed_Design_3',
        content: '<xml><variables></variables></xml>',
        user: {
          username: 'admin',
        },
      }, {
        id: 5,
        name: 'Unnamed_Design_2',
        content: '<xml><variables></variables></xml>',
        user: {
          username: 'testuser',
        },
      }],
    };
    const wrapper = shallowWithIntl(
      <ProgramList
        programs={programs}
        userPrograms={programs}
        fetchPrograms={fetchPrograms}
        removeProgram={removeProgram}
        fetchTags={fetchTags}
        clearProgram={clearProgram}
        clearProgramList={clearProgramList}
        fetchUserStats={fetchUserStats}
        user={{ user_id: 1, username: 'testuser', tier: 1 }}
        owned
      />,
    ).dive().dive();

    expect(wrapper.find(ProgramCollection).exists()).toBe(true);
    expect(wrapper.find(CircularProgress).exists()).toBe(false);
  });

  test('shows loading when programs fetching', () => {
    const wrapper = shallowWithIntl(
      <ProgramList
        fetchPrograms={fetchPrograms}
        removeProgram={removeProgram}
        fetchTags={fetchTags}
        clearProgram={clearProgram}
        clearProgramList={clearProgramList}
        fetchUserStats={fetchUserStats}
        user={{ user_id: 1, username: 'testuser', tier: 1 }}
        programs={null}
        owned
      />,
    ).dive().dive();

    expect(wrapper.find(ProgramCollection).exists()).toBe(false);
    expect(wrapper.find(CircularProgress).exists()).toBe(true);
    expect(wrapper.find(CircularProgress).length).toBe(1);
  });

  test('redirects to mission control when program loads', () => {
    const wrapper = shallowWithIntl(
      <ProgramList
        fetchPrograms={fetchPrograms}
        removeProgram={removeProgram}
        fetchTags={fetchTags}
        clearProgram={clearProgram}
        clearProgramList={clearProgramList}
        fetchUserStats={fetchUserStats}
        user={{ user_id: 1, username: 'testuser', tier: 1 }}
        owned
      />,
    ).dive().dive();

    wrapper.setState({
      programSelected: 1,
    });

    expect(wrapper.find(Redirect).exists()).toBe(true);
    expect(wrapper.find(Redirect).at(0).prop('to')).toEqual({
      pathname: '/mission-control/1',
    });
  });

  test('loads a program', () => {
    const programs = {
      next: null,
      previous: null,
      total_pages: 1,
      results: [{
        id: 33,
        name: 'Unnamed_Design_3',
        content: '<xml><variables></variables></xml>',
        user: {
          username: 'admin',
        },
      }],
    };
    const wrapper = shallowWithIntl(
      <ProgramList
        programs={programs}
        fetchPrograms={fetchPrograms}
        removeProgram={removeProgram}
        fetchTags={fetchTags}
        clearProgram={clearProgram}
        clearProgramList={clearProgramList}
        fetchUserStats={fetchUserStats}
        user={{ user_id: 1, username: 'testuser', tier: 1 }}
        owned
      />,
    ).dive().dive();

    wrapper.instance().selectProgram({
      target: {
        parentNode: {
          parentNode: {
            id: undefined,
          },
        },
        id: 33,
        dataset: {
          owned: 'false',
        },
      },
    });
    expect(wrapper.state('programSelected')).toBe(33);

    wrapper.setState({
      programLoaded: false,
    });

    wrapper.instance().selectProgram({
      target: {
        parentNode: {
          parentNode: {
            id: 55,
            dataset: {
              owned: 'true',
            },
          },
        },
      },
    });
    expect(wrapper.state('programSelected')).toBe(55);
  });

  test('fetches user programs after page change', () => {
    const programs = {
      next: null,
      previous: null,
      total_pages: 1,
      results: [{
        id: 33,
        name: 'Unnamed_Design_3',
        content: '<xml><variables></variables></xml>',
        user: {
          username: 'admin',
        },
      }],
    };
    const wrapper = shallowWithIntl(
      <ProgramList
        programs={programs}
        fetchPrograms={fetchPrograms}
        removeProgram={removeProgram}
        fetchTags={fetchTags}
        clearProgram={clearProgram}
        clearProgramList={clearProgramList}
        fetchUserStats={fetchUserStats}
        user={{ user_id: 1, username: 'testuser', tier: 1 }}
        owned
      />,
    ).dive().dive();

    wrapper.instance().fetchUserPrograms({
      page: 2,
    }, true);

    expect(fetchPrograms).toHaveBeenCalledWith({
      user: 1,
      page: 2,
    });
  });

  test('fetches community programs after page change', () => {
    const programs = {
      next: null,
      previous: null,
      total_pages: 1,
      results: [{
        id: 33,
        name: 'Unnamed_Design_3',
        content: '<xml><variables></variables></xml>',
        user: {
          username: 'admin',
        },
      }],
    };
    const wrapper = shallowWithIntl(
      <ProgramList
        programs={programs}
        fetchPrograms={fetchPrograms}
        removeProgram={removeProgram}
        fetchTags={fetchTags}
        clearProgram={clearProgram}
        clearProgramList={clearProgramList}
        fetchUserStats={fetchUserStats}
        user={{ user_id: 1, username: 'testuser', tier: 1 }}
        owned={false}
      />,
    ).dive().dive();

    wrapper.instance().fetchCommunityPrograms({
      page: 2,
    }, false);

    expect(fetchPrograms).toHaveBeenCalledWith({
      user__not: 1,
      page: 2,
    });
  });

  test('fetches programs after search change', () => {
    const programs = {
      next: null,
      previous: null,
      total_pages: 1,
      results: [{
        id: 33,
        name: 'Unnamed_Design_3',
        content: '<xml><variables></variables></xml>',
        user: {
          username: 'admin',
        },
      }],
    };
    const wrapper = shallowWithIntl(
      <ProgramList
        programs={programs}
        fetchPrograms={fetchPrograms}
        removeProgram={removeProgram}
        fetchTags={fetchTags}
        clearProgram={clearProgram}
        clearProgramList={clearProgramList}
        fetchUserStats={fetchUserStats}
        user={{ user_id: 1, username: 'testuser', tier: 1 }}
        owned
      />,
    ).dive().dive();

    wrapper.instance().fetchUserPrograms({
      search: 'abc',
      page: 1,
    }, true);

    expect(fetchPrograms).toHaveBeenCalledWith({
      user: 1,
      page: 1,
      search: 'abc',
    });
  });

  test('fetches other programs after search change', () => {
    const programs = {
      next: null,
      previous: null,
      total_pages: 1,
      results: [{
        id: 33,
        name: 'Unnamed_Design_3',
        content: '<xml><variables></variables></xml>',
        user: {
          username: 'admin',
        },
      }],
    };
    const wrapper = shallowWithIntl(
      <ProgramList
        programs={programs}
        fetchPrograms={fetchPrograms}
        removeProgram={removeProgram}
        fetchTags={fetchTags}
        clearProgram={clearProgram}
        clearProgramList={clearProgramList}
        fetchUserStats={fetchUserStats}
        user={{ user_id: 1, username: 'testuser', tier: 1 }}
        owned={false}
      />,
    ).dive().dive();

    wrapper.instance().fetchCommunityPrograms({
      search: 'abc',
      page: 1,
    }, false);

    expect(fetchPrograms).toHaveBeenCalledWith({
      user__not: 1,
      page: 1,
      search: 'abc',
    });
  });

  test('removes a program and reloads the program list', (done) => {
    const programs = {
      next: null,
      previous: null,
      total_pages: 1,
      results: [{
        id: 33,
        name: 'Unnamed_Design_3',
        content: '<xml><variables></variables></xml>',
        user: {
          username: 'admin',
        },
      }],
    };
    const wrapper = shallowWithIntl(
      <ProgramList
        programs={programs}
        fetchPrograms={fetchPrograms}
        removeProgram={removeProgram}
        fetchTags={fetchTags}
        clearProgram={clearProgram}
        clearProgramList={clearProgramList}
        fetchUserStats={fetchUserStats}
        user={{ user_id: 1, username: 'testuser', tier: 1 }}
        owned
      />,
    ).dive().dive();

    wrapper.setState({
      focusProgram: {
        id: 33,
        name: 'Unnamed_Design_3',
      },
    });
    wrapper.instance().removeProgram().then(() => {
      expect(fetchPrograms).toHaveBeenCalledTimes(2);
      expect(removeProgram).toHaveBeenCalledWith(33);
      done();
    });
  });

  test('shows confirm dialog', () => {
    const wrapper = shallowWithIntl(
      <ProgramList
        fetchPrograms={fetchPrograms}
        removeProgram={removeProgram}
        fetchTags={fetchTags}
        clearProgram={clearProgram}
        clearProgramList={clearProgramList}
        fetchUserStats={fetchUserStats}
        user={{ user_id: 1, username: 'testuser', tier: 1 }}
        owned
      />,
    ).dive().dive();

    wrapper.instance().showConfirm({
      target: {
        parentNode: {
          parentNode: {
            id: undefined,
          },
        },
        id: 33,
        name: 'Unnamed_Design_3',
      },
    });

    expect(wrapper.state('confirmOpen')).toBe(true);
  });

  test('cancel dialog does not remove program', () => {
    const wrapper = shallowWithIntl(
      <ProgramList
        fetchPrograms={fetchPrograms}
        removeProgram={removeProgram}
        fetchTags={fetchTags}
        clearProgram={clearProgram}
        clearProgramList={clearProgramList}
        fetchUserStats={fetchUserStats}
        user={{ user_id: 1, username: 'testuser', tier: 1 }}
        owned
      />,
    ).dive().dive();

    wrapper.instance().cancelRemove();

    expect(fetchPrograms).toHaveBeenCalledTimes(1);
    expect(removeProgram).not.toHaveBeenCalled();
    expect(wrapper.state('confirmOpen')).toBe(false);
  });
});
