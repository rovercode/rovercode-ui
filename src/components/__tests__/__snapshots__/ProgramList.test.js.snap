// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The ProgramList component renders on the page with no errors 1`] = `
<Fragment>
  <WithStyles(ForwardRef(Grid))
    alignItems="center"
    container={true}
    direction="column"
    justify="center"
  >
    <WithStyles(ForwardRef(Grid))
      item={true}
    >
      <Styled(MuiBox)
        m={2}
      >
        <injectIntl(WithStyles(ProgramCollection))
          label="My Programs"
          onProgramClick={[Function]}
          onRemoveClick={[Function]}
          onUpdate={[Function]}
          programs={
            Object {
              "next": null,
              "previous": null,
              "results": Array [],
              "total_pages": 1,
            }
          }
          tag={
            Object {
              "tags": Array [],
            }
          }
          user={
            Object {
              "tier": 1,
              "user_id": 1,
              "username": "testuser",
            }
          }
        />
      </Styled(MuiBox)>
    </WithStyles(ForwardRef(Grid))>
  </WithStyles(ForwardRef(Grid))>
  <ConfirmDialog
    cancelButton="No, don't remove program"
    confirmButton="Yes, remove program"
    onCancel={[Function]}
    onConfirm={[Function]}
    open={false}
    title="Remove Program"
  >
    Are you sure you want to remove Test Program?
  </ConfirmDialog>
  <WithStyles(ForwardRef(Divider)) />
  <Footer />
</Fragment>
`;
